wikimedia/wikipedia-ios
# Notifications - Rework data flow & add ability to locally page

**Phabricator:** https://phabricator.wikimedia.org/T288662

### Notes
This PR modifies the data handling of Notification Center. It refactors out `NSFetchedResultsController` into basic `NSFetchRequests` in `RemoteNotificationsController`, and uses a diffable data source to animate the changes instead of `CollectionViewUpdater`. It also adds paging through local results and a basic edit mode to see the different cell styles.

I tested the performance and I didn't notice any issues with this particular code. I found other areas of improvement, but I will handle that in a followup PR.

### Test Steps
1. Fresh install app. Login to an account with notifications. Go to Settings > Languages and add additional app languages to get the number of notifications pretty large. You might want to consider tweaking the fetch limit [here](https://github.com/wikimedia/wikipedia-ios/blob/ba6d5225bbac52f533f30d221d86b1120e8a4920/WMF%20Framework/Remote%20Notifications/RemoteNotificationsController.swift#L31) to something lower if you have a smaller number of notifications so you can experience local paging.
2. Go to notification center. Wait for it to complete importing, then scroll around. Scroll down to the bottom, confirm next 50 cells are fetched and added to the list.
3. Tap the Edit button at the top, the view will switch to edit mode. Tap the circles in the cells and confirm they show as checked. Scroll down, then scroll back up. Confirm your checked cells are still checked.

### FYI
I've been working ahead in a `T288662-paging-future` prototype branch, if you're curious to see the data flow for future functionality. It has a pull to refresh, mark as read (just tapping the cell background for this), filtering (just one type), and 10,000 mock notifications importing for testing performance.
